File: src/main/java/com/jloom/registry/CatalogRoots.java
package com.jloom.registry;

import java.io.InputStream;
import java.util.List;

public final class CatalogRoots {

    private static final List<String> ALL = List.of("/modules/", "/services/");

    private CatalogRoots() {
    }

    public static InputStream open(String moduleId, String relativePath) {
        for (String root : ALL) {
            InputStream in = CatalogRoots.class.getResourceAsStream(root + moduleId + "/" + relativePath);
            if (in != null) {
                return in;
            }
        }
        return null;
    }

    static List<String> indexResources() {
        return ALL.stream().map(root -> root + "modules.yml").toList();
    }
}

File: node/src/registry/catalogRoots.ts
import { existsSync, readFileSync } from "node:fs";
import path from "node:path";

// Mirrors com.jloom.registry.CatalogRoots: two catalog roots, tried in order.
const ROOTS = ["modules", "services"];

const CATALOG_ROOT = path.join(import.meta.dirname, "..", "..", "catalog");

export function catalogRoot(): string {
  return CATALOG_ROOT;
}

export function resolvePath(id: string, relativePath: string): string | undefined {
  for (const root of ROOTS) {
    const candidate = path.join(CATALOG_ROOT, root, id, relativePath);
    if (existsSync(candidate)) {
      return candidate;
    }
  }
  return undefined;
}

export function readText(id: string, relativePath: string): string | undefined {
  const resolved = resolvePath(id, relativePath);
  return resolved === undefined ? undefined : readFileSync(resolved, "utf8");
}

export function indexPaths(): string[] {
  return ROOTS.map((root) => path.join(CATALOG_ROOT, root, "modules.yml"));
}
